tests/test_semdedup.py

```diff
@@ -233,3 +245,57 @@ def get_reference_embeddings(
         embs.append(normed_emb)
 
     return np.array(embs)
+
+
+class TestPairwiseCosineSimilarity:
+    def setup_method(self):
+        # We create a 5x3 array where each row is a unit vector
+        # The second and last two rows are the same
+        input_arr = torch.tensor(
+            np.asarray(
+                [[1, 2, 3], [4, 5, 6], [7, 8, 9], [10, 11, 12], [1, 2, 3], [1, 2, 3]],
+            ),
+            dtype=torch.float32,
+        )
+        # Normalize the input array
+        self.input_arr = input_arr / torch.norm(input_arr, dim=1, keepdim=True)
+        self.expected_similarity = torch.tensor(
+            [0.0000, 0.974631, 0.998190, 0.999618, 1.0000, 1.0000]
+        )
+        self.expected_indices = [0, 0, 1, 2, 0, 0]
+
+    @pytest.mark.parametrize("device", [pytest.param("cuda", marks=pytest.mark.gpu)])
+    def test_pairwise_cosine_similarity(self, device: Literal["cpu", "cuda"]):
+        max_similarity, max_indices = pairwise_cosine_similarity(
+            self.input_arr.to(device), device
+        )
+        torch.testing.assert_close(
+            max_similarity, self.expected_similarity, rtol=1e-6, atol=1e-6
+        )
+        assert max_indices == self.expected_indices
+
+    @pytest.mark.parametrize("device", [pytest.param("cuda", marks=pytest.mark.gpu)])
+    @pytest.mark.parametrize("batch_size", [1, 2, 3, 4, 5, 6])
+    def test_pairwise_cosine_similarity_batched(
+        self, device: Literal["cpu", "cuda"], batch_size: int
+    ):
+        max_similarity, max_indices = pairwise_cosine_similarity_batched(
+            self.input_arr.to(device), device, batch_size
+        )
+        torch.testing.assert_close(max_similarity, self.expected_similarity)
+        assert max_indices == self.expected_indices
+
+    @pytest.mark.parametrize("device", [pytest.param("cuda", marks=pytest.mark.gpu)])
+    @pytest.mark.parametrize("batch_size", [100, 512, 1024, 2048])
+    def test_pairwise_cosine_similarity_batched_rand_array(
+        self, device: Literal["cpu", "cuda"], batch_size: int
+    ):
+        N = 1024
+        D = 512
+        rand_arr = torch.randn(N, D, device=device)
+        max_similarity, max_indices = pairwise_cosine_similarity(rand_arr, device)
+        max_similarity_batched, max_indices_batched = (
+            pairwise_cosine_similarity_batched(rand_arr, device, batch_size=batch_size)
+        )
+        torch.testing.assert_close(max_similarity, max_similarity_batched)
```

Comment:
@ryantwolf even for this where we thought assert equality should be fine, even then it wasn't. Long live floating imprecisions

E       Mismatched elements: 1 / 1024 (0.1%)
E       Greatest absolute difference: 1.33514404296875e-05 at index (3,) (up to 1e-05 allowed)
E       Greatest relative difference: 6.5386520873289555e-06 at index (3,) (up to 1.3e-06 allowed)